File: server/services/availability.js
/* Centralized pandit availability engine (master plan Phase 3).
   ONE module answers "can this pandit take this booking?" for every path:
   customer booking, auto-assignment, admin assignment, rescheduling.

   Check order (first failure wins, each with a human-readable reason):
     1  pandit exists
     2  KYC status verified
     3  accepting bookings (avail flag)
     4  weekly off            (weekly_off, JSON array of 0=Sun..6=Sat)
     5  holiday               (holidays, JSON array of ISO dates — recurring meaning)
     6  blocked date          (blocked_dates, JSON array of ISO dates + reason)
     7  marked-off date       (existing pandits.off — legacy single-date off)
     8  time slot             (slots, JSON array; empty = all shared SLOTS)
     9  online capability     (online_enabled)
     10 temple capability     (temple_enabled)
     11 home-puja radius      (radius_km + base coords; NULL = not enforced)
     12 slot conflict         (existing partial unique index is the last line of defense)

   Defaults are deliberately permissive: empty arrays / NULLs mean "no restriction",
   so existing pandits and seeded data behave exactly as before (Phase 36). */
'use strict';
const { db } = require('../db');
const { j } = require('../lib/util');
const P = require('../../shared/pricing');

const WD = ['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday'];

/* Haversine distance in km (never imported — one tiny local helper). */
function haversineKm(lat1, lon1, lat2, lon2) {
  const R = 6371, toRad = (d) => (d * Math.PI) / 180;
  const dLat = toRad(lat2 - lat1), dLon = toRad(lon2 - lon1);
  const a = Math.sin(dLat / 2) ** 2 + Math.cos(toRad(lat1)) * Math.cos(toRad(lat2)) * Math.sin(dLon / 2) ** 2;
  return 2 * R * Math.asin(Math.sqrt(a));
}

/* Resolve a booking location (from a booking row or a request body) to coords.
   Tolerant city match: 'Delhi NCR' matches place_index 'Delhi'. */
function resolvePlace(city) {
  if (!city) return null;
  const row = db.prepare(`SELECT lat, lon FROM place_index WHERE country='India' AND
    (city = ? OR ? LIKE city || ' %' OR city LIKE ? || ' %') ORDER BY population DESC LIMIT 1`)
    .get(city, city, city);
  return row || null;
}

/* The bookable formula. mode may be undefined for legacy checks (slot-only). */
function check(p, date, slot, { mode, city, lat, lon, skipId } = {}) {
  if (!p) return { ok: false, reason: 'Pandit not found', code: 'NOT_FOUND' };
  if (p.status !== 'verified') return { ok: false, reason: 'KYC verification pending', code: 'KYC' };
  if (!p.avail) return { ok: false, reason: 'Pandit is not accepting new bookings', code: 'AVAIL_FLAG' };

  const weekly = j(p.weekly_off, []);
  const wd = new Date(date + 'T12:00:00Z').getUTCDay();
  if (weekly.includes(wd)) return { ok: false, reason: WD[wd] + ' is a weekly off for this pandit', code: 'WEEKLY_OFF' };

  const holidays = j(p.holidays, []);
  if (holidays.includes(date)) return { ok: false, reason: 'Pandit is on holiday on this date', code: 'HOLIDAY' };

  const blocked = j(p.blocked_dates, []);
  const hit = blocked.find((b) => (b && b.date) === date);
  if (hit) return { ok: false, reason: 'Date blocked by the pandit' + (hit.reason ? ': ' + hit.reason : ''), code: 'BLOCKED' };

  if (j(p.off, []).includes(date)) return { ok: false, reason: 'Pandit marked this date unavailable', code: 'MARKED_OFF' };

  const configured = j(p.slots, []);
  if (slot && configured.length && !configured.includes(slot)) {
    return { ok: false, reason: 'Pandit does not take bookings in the ' + slot + ' slot', code: 'SLOT' };
  }

  if (mode === 'online' && !p.online_enabled) return { ok: false, reason: 'Pandit does not offer online pujas', code: 'ONLINE_OFF' };
  if (mode === 'temple' && !p.temple_enabled) return { ok: false, reason: 'Pandit does not offer temple services', code: 'TEMPLE_OFF' };

  if (mode === 'home' && p.radius_km != null) {
    const base = (p.base_lat != null && p.base_lon != null)
      ? { lat: p.base_lat, lon: p.base_lon }
      : resolvePlace(p.city);
    if (!base) return { ok: false, reason: 'Pandit base location is unknown; radius cannot be checked', code: 'RADIUS' };
    let dest = (lat != null && lon != null) ? { lat: Number(lat), lon: Number(lon) } : null;
    if (!dest) {
      const reqCity = city || (() => { const b = db.prepare("SELECT addr FROM bookings WHERE id=?").get(skipId || ''); return b ? (j(b.addr, {}).city) : null; })();
      dest = resolvePlace(reqCity);
    }
    if (!dest) return { ok: false, reason: 'Location is outside the pandit\u2019s service area', code: 'RADIUS' };
    const km = haversineKm(base.lat, base.lon, dest.lat, dest.lon);
    if (km > p.radius_km) {
      return { ok: false, reason: 'Location is ' + Math.round(km) + ' km away, beyond the pandit\u2019s ' + p.radius_km + ' km service radius', code: 'RADIUS' };
    }
  }

  /* Slot conflict LAST: the DB partial unique index remains the hard guarantee. */
  if (slot) {
    const c = db.prepare("SELECT 1 FROM bookings WHERE pandit_id=? AND date=? AND slot=? AND status NOT IN ('Cancelled') AND id != ?")
      .get(p.id, date, slot, skipId || '');
    if (c) return { ok: false, reason: 'Already booked in the ' + slot + ' slot on this date', code: 'CONFLICT' };
  }
  return { ok: true, reason: 'Available', code: 'OK' };
}

/* Boolean twin of check() for drop-in use at legacy call sites. */
function isFree(p, date, slot, opts = {}) { return check(p, date, slot, opts).ok; }

/* Best-pandit picker honoring availability, then spec/city/rating as before. */
function autoPick(pujaId, city, date, slot, { mode, lat, lon } = {}) {
  const list = db.prepare("SELECT * FROM pandits WHERE status='verified'").all()
    .filter((p) => isFree(p, date, slot, { mode, city, lat, lon }));
  list.sort((a, b) => (j(b.spec, []).includes(pujaId) - j(a.spec, []).includes(pujaId))
    || ((b.city === city) - (a.city === city)) || b.rating - a.rating);
  return list[0] || null;
}

/* Which pandits can take (puja, mode, date, slot)? Powers the customer wizard's
   "available pandits" UX. Returns rows with a machine + human reason each. */
function whoIsAvailable(pujaId, city, date, slot, { mode, lat, lon } = {}) {
  return db.prepare("SELECT * FROM pandits WHERE status='verified'").all()
    .map((p) => ({ p, r: check(p, date, slot, { mode, city, lat, lon }) }))
    .filter((x) => x.r.ok)
    .sort((a, b) => (j(b.p.spec, []).includes(pujaId) - j(a.p.spec, []).includes(pujaId))
      || ((b.p.city === city) - (a.p.city === city)) || b.p.rating - a.p.rating)
    .map((x) => x.p);
}

/* Availability configuration for the pandit portal / admin screens. */
function configOf(p) {
  return { weeklyOff: j(p.weekly_off, []), slots: j(p.slots, []), holidays: j(p.holidays, []),
           blockedDates: j(p.blocked_dates, []), radiusKm: p.radius_km != null ? p.radius_km : null,
           baseLat: p.base_lat, baseLon: p.base_lon,
           onlineEnabled: !!p.online_enabled, templeEnabled: !!p.temple_enabled };
}

module.exports = { check, isFree, autoPick, whoIsAvailable, resolvePlace, haversineKm, configOf };

File: backend-python/app/services/availability.py
"""Centralized pandit availability engine (Phase 3) — twin of
server/services/availability.js. ONE module answers "can this pandit take this
booking?" for every path: customer booking, auto-assignment, admin assignment,
rescheduling. Check order and permissive defaults match the Node twin exactly
(empty lists / NULL radius = no restriction, so existing data keeps working)."""
import math

from sqlalchemy import select

from ..models import Booking, Pandit, PlaceIndex
from ..util import j

WD = ["Sunday", "Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday"]


def _js_date_wd(iso_date: str) -> int:
    """JS getUTCDay() parity: 0=Sunday..6=Saturday, noon-anchored like the Node twin."""
    from datetime import datetime

    d = datetime.fromisoformat(iso_date + "T12:00:00+00:00")
    return (d.weekday() + 1) % 7  # Python Monday=0 -> JS Sunday=0 mapping


def haversine_km(lat1: float, lon1: float, lat2: float, lon2: float) -> float:
    r = 6371.0
    to_rad = math.pi / 180
    dlat = (lat2 - lat1) * to_rad
    dlon = (lon2 - lon1) * to_rad
    a = math.sin(dlat / 2) ** 2 + math.cos(lat1 * to_rad) * math.cos(lat2 * to_rad) * math.sin(dlon / 2) ** 2
    return 2 * r * math.asin(math.sqrt(a))


async def resolve_place(db, city: str | None):
    """Tolerant city match: 'Delhi NCR' matches place_index 'Delhi'."""
    if not city:
        return None
    rows = (await db.execute(
        select(PlaceIndex).where(PlaceIndex.country == "India")
        .order_by(PlaceIndex.population.desc()))).scalars().all()
    c = str(city)
    for row in rows:
        if row.city == c or c.startswith(row.city + " ") or row.city.startswith(c + " "):
            return row
    return None


async def check(db, p: Pandit | None, date: str, slot: str | None, *,
                mode: str | None = None, city: str | None = None,
                lat=None, lon=None, skip_id: str = "") -> dict:
    """The bookable formula. Returns {ok, reason, code} — first failure wins."""
    if not p:
        return {"ok": False, "reason": "Pandit not found", "code": "NOT_FOUND"}
    if p.status != "verified":
        return {"ok": False, "reason": "KYC verification pending", "code": "KYC"}
    if not p.avail:
        return {"ok": False, "reason": "Pandit is not accepting new bookings", "code": "AVAIL_FLAG"}

    weekly = j(p.weekly_off, [])
    wd = _js_date_wd(date)
    if wd in weekly:
        return {"ok": False, "reason": WD[wd] + " is a weekly off for this pandit", "code": "WEEKLY_OFF"}

    if date in j(p.holidays, []):
        return {"ok": False, "reason": "Pandit is on holiday on this date", "code": "HOLIDAY"}

    blocked = j(p.blocked_dates, [])
    hit = next((b for b in blocked if isinstance(b, dict) and b.get("date") == date), None)
    if hit:
        return {"ok": False, "reason": "Date blocked by the pandit" + (": " + hit["reason"] if hit.get("reason") else ""), "code": "BLOCKED"}

    if date in j(p.off, []):
        return {"ok": False, "reason": "Pandit marked this date unavailable", "code": "MARKED_OFF"}

    configured = j(p.slots, [])
    if slot and configured and slot not in configured:
        return {"ok": False, "reason": f"Pandit does not take bookings in the {slot} slot", "code": "SLOT"}

    if mode == "online" and not p.online_enabled:
        return {"ok": False, "reason": "Pandit does not offer online pujas", "code": "ONLINE_OFF"}
    if mode == "temple" and not p.temple_enabled:
        return {"ok": False, "reason": "Pandit does not offer temple services", "code": "TEMPLE_OFF"}

    if mode == "home" and p.radius_km is not None:
        base = None
        if p.base_lat is not None and p.base_lon is not None:
            base = (p.base_lat, p.base_lon)
        else:
            pl = await resolve_place(db, p.city)
            base = (pl.lat, pl.lon) if pl else None
        if not base:
            return {"ok": False, "reason": "Pandit base location is unknown; radius cannot be checked", "code": "RADIUS"}
        if lat is not None and lon is not None:
            dest = (float(lat), float(lon))
        else:
            pl = await resolve_place(db, city)
            dest = (pl.lat, pl.lon) if pl else None
        if not dest:
            return {"ok": False, "reason": "Location is outside the pandit\u2019s service area", "code": "RADIUS"}
        km = haversine_km(base[0], base[1], dest[0], dest[1])
        if km > p.radius_km:
            return {"ok": False, "reason": f"Location is {round(km)} km away, beyond the pandit\u2019s {p.radius_km} km service radius", "code": "RADIUS"}

    # Slot conflict LAST: the DB partial unique index remains the hard guarantee.
    if slot:
        clash = (await db.execute(
            select(Booking.id).where(
                Booking.pandit_id == p.id, Booking.date == date, Booking.slot == slot,
                Booking.status != "Cancelled", Booking.id != skip_id).limit(1))).scalar_one_or_none()
        if clash is not None:
            return {"ok": False, "reason": f"Already booked in the {slot} slot on this date", "code": "CONFLICT"}
    return {"ok": True, "reason": "Available", "code": "OK"}


async def is_free(db, p, date: str, slot: str, **kw) -> bool:
    return (await check(db, p, date, slot, **kw))["ok"]


async def auto_pick(db, puja_id: str, city: str | None, date: str, slot: str, *,
                    mode: str | None = None, lat=None, lon=None) -> Pandit | None:
    rows = (await db.execute(select(Pandit).where(Pandit.status == "verified"))).scalars().all()
    free = [p for p in rows if (await check(db, p, date, slot, mode=mode, city=city, lat=lat, lon=lon))["ok"]]

    def score(p: Pandit):
        return (puja_id in j(p.spec, []), p.city == city, p.rating or 0)
    free.sort(key=score, reverse=True)
    return free[0] if free else None


async def who_is_available(db, puja_id: str, city: str | None, date: str, slot: str, *,
                           mode: str | None = None, lat=None, lon=None) -> list[Pandit]:
    rows = (await db.execute(select(Pandit).where(Pandit.status == "verified"))).scalars().all()
    ok = [p for p in rows if (await check(db, p, date, slot, mode=mode, city=city, lat=lat, lon=lon))["ok"]]

    def score(p: Pandit):
        return (puja_id in j(p.spec, []), p.city == city, p.rating or 0)
    ok.sort(key=score, reverse=True)
    return ok


def config_of(p: Pandit) -> dict:
    """Availability configuration for the pandit portal / admin screens."""
    return {"weeklyOff": j(p.weekly_off, []), "slots": j(p.slots, []),
            "holidays": j(p.holidays, []), "blockedDates": j(p.blocked_dates, []),
            "radiusKm": p.radius_km, "baseLat": p.base_lat, "baseLon": p.base_lon,
            "onlineEnabled": bool(p.online_enabled), "templeEnabled": bool(p.temple_enabled)}
